core/trino-main/src/main/java/io/trino/operator/output/BytePositionsAppender.java

```diff
@@ -14,31 +14,239 @@
 package io.trino.operator.output;
 
 import io.trino.spi.block.Block;
-import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.BlockBuilderStatus;
+import io.trino.spi.block.BlockUtil;
+import io.trino.spi.block.ByteArrayBlock;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.trino.spi.block.BlockUtil.calculateBlockResetSize;
+import static java.lang.Math.max;
 
 public class BytePositionsAppender
-        implements PositionsAppender
+        implements BlockTypeAwarePositionsAppender
 {
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BytePositionsAppender.class).instanceSize();
+    private static final Block NULL_VALUE_BLOCK = new ByteArrayBlock(1, Optional.of(new boolean[] {true}), new byte[1]);
+
+    @Nullable
+    private final BlockBuilderStatus blockBuilderStatus;
+    private boolean initialized;
+    private final int initialEntryCount;
+
+    private int positionCount;
+    private boolean hasNullValue;
+    private boolean hasNonNullValue;
+
+    // it is assumed that these arrays are the same length
+    private boolean[] valueIsNull = new boolean[0];
+    private byte[] values = new byte[0];
+
+    private long retainedSizeInBytes;
+
+    public BytePositionsAppender(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries)
```

Comment:
This is simply clean code. Just because not all codebase follow it does not mean you shouldn't.
The idea is simple: If there are only a few possible combinations (like here) - make all of them explicit. If there are more, make a builder.

Reply:
At the very least make the nullable argument the last one.

Reply:
clean code matters but consistency in the codebase also matters.
I wanted PossitionAppender to be a drop-in replacement for BlockBuilder but maybe that is too far out for me to worry about it.
I'm gonna refactor to 2 constructors but after we decide what to do with  `BlockBuilderStatus`(it may be dropped anyway).